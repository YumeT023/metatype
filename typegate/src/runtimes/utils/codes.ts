export interface Code {
  name: string;
  source: string;
  type: "module" | "func";
}

export interface TaskContext {
  parent?: Record<string, unknown>;
  context?: Record<string, string>;
}

export interface FunctionMaterializerData {
  fn_expr: string;
}

export interface ImportFuncMaterializerData {
  mod: number;
  name: string;
}

interface TaskBase {
  type: "func" | "import_func" | "predefined_func";
  id: number;
  args: Record<string, unknown>;
  context: TaskContext;
  verbose: boolean;
}

export interface ImportFuncTask extends TaskBase {
  type: "import_func";
  moduleId: number;
  moduleCode?: string;
  name: string;
}

export interface FuncTask extends TaskBase {
  type: "func";
  fnId: number;
  code?: string;
}

export interface PredefinedFuncTask extends TaskBase {
  type: "predefined_func";
  name: string;
}

export type Task = ImportFuncTask | FuncTask | PredefinedFuncTask;

export interface TaskExec {
  (args: Record<string, unknown>, context: TaskContext): unknown;
}

export const predefinedFuncs: Record<string, TaskExec> = {
  identity: (args) => args,
};
