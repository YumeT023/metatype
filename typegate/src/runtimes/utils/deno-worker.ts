import { getLogger } from "../../log.ts";
import {
  FuncTask,
  ImportFuncTask,
  predefinedFuncs,
  PredefinedFuncTask,
  Task,
  TaskExec,
} from "./codes.ts";

let logger = getLogger("worker");
logger.info("start webworker");

let initData = null as unknown as { name: string };

type TaskModule = Record<string, TaskExec>;

const fns: Map<number, TaskExec> = new Map();
const mods: Map<number, TaskModule> = new Map();

// TODO: get worker name from events, for better logging

const execFunctions: Record<Task["type"], (task: Task) => Promise<unknown>> = {
  import_func: async (task: Task) => {
    const { id, moduleId, moduleCode, name, args, context, verbose } =
      task as ImportFuncTask;
    if (!mods.has(moduleId)) {
      if (moduleCode == null) {
        throw new Error("module definition required for first reference");
      }
      mods.set(
        moduleId,
        await import(
          `data:text/javascript,${encodeURIComponent(moduleCode)}`
        ),
      );
    }

    verbose &&
      logger.info(`[${id}] exec func "${name}" from module ${moduleId}`);
    const mod = mods.get(moduleId)!;
    return await mod[name](args, context);
  },

  func: async (task: Task) => {
    const { id, fnId, code, args, context, verbose } = task as FuncTask;
    if (!fns.has(fnId)) {
      if (code == null) {
        throw new Error("function definition required");
      }
      fns.set(fnId, new Function(`"use strict"; return ${code}`)());
    }

    verbose && logger.info(`[${id}] exec func "${fnId}"`);
    return await fns.get(fnId)!(args, context);
  },

  predefined_func: (task: Task) => {
    const { id, name, args, context, verbose } = task as PredefinedFuncTask;
    verbose && logger.info(`[${id}] exec predefined func "${name}"`);
    return Promise.resolve(predefinedFuncs[name](args, context));
  },
};

self.onmessage = async (evt: MessageEvent<Task>) => {
  if (initData == null) {
    initData = evt.data as typeof initData;
    logger = getLogger(`worker (${initData.name})`);
    return;
  }

  const { id, type } = evt.data;
  const exec = execFunctions[type];

  if (exec == null) {
    self.postMessage({
      id,
      error: `unsupported task type "${evt.data.type}"`,
    });
  }

  try {
    const value = await exec(evt.data);
    self.postMessage({ id, value });
  } catch (err) {
    self.postMessage({ id, error: err.message });
  }
};
