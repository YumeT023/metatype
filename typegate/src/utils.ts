import { ComputeStage } from "./engine.ts";
import * as ast from "graphql_ast";

// FIXME replace with monads
export type Maybe<T> = null | undefined | T;
export type JSONValue =
  | string
  | number
  | boolean
  | null
  | JSONValue[]
  | { [key: string]: JSONValue };

export const ensure = (predicat: boolean, message: string | (() => string)) => {
  if (!predicat) {
    throw Error(typeof message === "function" ? message() : message);
  }
};

export const collectFields = (
  obj: Record<string, unknown>,
  fields: string[],
) => {
  return fields.reduce((agg, f) => ({ ...agg, [f]: obj[f] }), {});
};

export const b = (value: any): string => JSON.stringify(value, null, 2);

// FIXME remplace all instance
export const mapo = <V1, V2>(
  vs: Record<string, V1>,
  map: (e: V1) => V2,
): Record<string, V2> =>
  Object.entries(vs).reduce((agg, [key, value]) => {
    agg[key] = map(value);
    return agg;
  }, {} as Record<string, V2>);

export const unparse = (loc: ast.Location): string => {
  return loc.source.body.slice(loc.start, loc.end);
};

export function iterParentStages(
  stages: ComputeStage[],
  cb: (stage: ComputeStage, children: ComputeStage[]) => void,
) {
  let cursor = 0;
  while (cursor < stages.length) {
    const stage = stages[cursor];
    const children = stages.slice(cursor + 1).filter((s) =>
      s.id().startsWith(stage.id())
    );
    cb(stage, children);
    cursor += 1 + children.length;
  }
}

export function unzip<A, B>(arrays: ([A, B])[]): [A[], B[]] {
  const as: A[] = [];
  const bs: B[] = [];
  arrays.forEach(([a, b]) => {
    as.push(a);
    bs.push(b);
  });
  return [as, bs];
}
