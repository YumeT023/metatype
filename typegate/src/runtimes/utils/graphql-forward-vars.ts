import { JSONValue, unparse, unzip } from "../../utils.ts";
import type { FromVars } from "../GraphQLRuntime.ts";
import { ComputeStage } from "../../engine.ts";
import * as ast from "graphql_ast";
import { iterParentStages } from "../../utils.ts";

export function stringifyQL(
  obj: JSONValue | FromVars<JSONValue>,
): [q: string, vars: string[]] {
  if (typeof obj === "function") {
    const varName = obj(null as unknown as Record<string, unknown>);
    // assert typeof varName === "string"
    return [`$${varName}`, [varName as string]];
  }

  if (Array.isArray(obj)) {
    const [qs, vars] = unzip(obj.map(stringifyQL));
    return [`[${qs.join(", ")}]`, vars.flat(1)];
  }

  if (typeof obj === "object" && obj != null) {
    const [qs, vars] = unzip(Object.entries(obj)!.map(([k, v]) => {
      const [q, vars] = stringifyQL(v);
      return [[k, q] as const, vars];
    }));
    return [
      `{${
        qs.map(([k, v]) => `${k}: ${v}`)
          .join(", ")
      }}`,
      vars.flat(1),
    ];
  }

  return [JSON.stringify(obj), []];
}

export interface RebuildQueryParam {
  stages: ComputeStage[];
  renames: Record<string, string>;
  varDefs: Record<string, ast.VariableDefinitionNode>;
}

export function rebuildGraphQuery(
  { stages, renames, varDefs }: RebuildQueryParam,
): [query: string, forwaredVars: Record<string, string>] {
  let query = "";
  const forwardedVars: Record<string, string> = {};
  const forwardVar = (name: string) => {
    if (!Object.hasOwnProperty.call(forwardedVars, name)) {
      forwardedVars[name] = unparse(varDefs[name].loc!);
    }
  };

  iterParentStages(stages, (stage, children) => {
    const field = stage.props.path[stage.props.path.length - 1];
    query += ` ${field !== stage.props.node ? field + ": " : ""}${
      renames[stage.props.node] ?? stage.props.node
    }`;
    if (Object.keys(stage.props.args).length > 0) {
      query += `(${
        Object.entries(stage.props.args).map(
          ([argName, argValue]) => {
            const val = argValue({}, null);
            if (typeof val === "function") {
              const varName = val(null);
              forwardVar(varName);
              return `${argName}: $${varName}`;
            }
            const [q, vars] = stringifyQL(val as JSONValue);
            vars.forEach(forwardVar);
            return `${argName}: ${q}`;
          },
        ).join(", ")
      })`;
    }
    if (children.length > 0) {
      const [q, vars] = rebuildGraphQuery({
        stages: children,
        renames,
        varDefs,
      });
      query += ` {${q} }`;
      Object.assign(forwardedVars, vars);
    }
  });
  return [query, forwardedVars];
}
