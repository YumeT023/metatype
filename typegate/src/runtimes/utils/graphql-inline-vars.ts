import { iterParentStages, JSONValue } from "../../utils.ts";
import type { FromVars } from "../GraphQLRuntime.ts";
import { ComputeStage } from "../../engine.ts";

export function stringifyQL(
  obj: JSONValue | FromVars<JSONValue>,
): FromVars<string> {
  if (typeof obj === "function") {
    return (vars) => stringifyQL(obj(vars))(vars);
  }

  if (Array.isArray(obj)) {
    return (vars) => `[${obj.map((obj) => stringifyQL(obj)(vars)).join(", ")}]`;
  }

  if (typeof obj === "object" && obj !== null) {
    return (vars) =>
      `{ ${
        Object.entries(obj).map(([k, v]) => `${k}: ${stringifyQL(v)(vars)}`)
      } }`;
  }

  return (_vars) => JSON.stringify(obj);
}

export interface RebuildQueryParam {
  stages: ComputeStage[];
  renames: Record<string, string>;
}

export function rebuildGraphQuery(
  { stages, renames }: RebuildQueryParam,
): FromVars<string> {
  const ss: Array<FromVars<string>> = [];

  iterParentStages(stages, (stage, children) => {
    const field = stage.props.path[stage.props.path.length - 1];
    ss.push(() =>
      ` ${field !== stage.props.node ? field + ": " : ""}${
        renames[stage.props.node] ?? stage.props.node
      }`
    );

    if (Object.keys(stage.props.args).length > 0) {
      ss.push((vars) =>
        `(${
          Object.entries(stage.props.args).map(([argName, argValue]) => {
            return `${argName}: ${
              stringifyQL(
                argValue({}, null, {}) as JSONValue | FromVars<JSONValue>,
              )(vars)
            }`;
          }).join(", ")
        })`
      );
    }

    if (children.length > 0) {
      ss.push((vars) =>
        ` {${rebuildGraphQuery({ stages: children, renames })(vars)} }`
      );
    }
  });

  return (vars) => ss.map((s) => s(vars)).join("");
}
