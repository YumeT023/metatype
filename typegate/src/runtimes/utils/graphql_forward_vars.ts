// Copyright Metatype OÃœ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

import { ensure, JSONValue, unzip } from "../../utils.ts";
import type { FromVars } from "../graphql.ts";
import { ComputeStage } from "../../engine.ts";
import { iterParentStages } from "../../utils.ts";
import {
  ArgumentNode,
  FieldNode,
  SelectionNode,
  VariableDefinitionNode,
} from "graphql/ast";
import { Kind, parseType } from "graphql";
import * as GraphQL from "graphql";

export function stringifyQL(
  obj: JSONValue | FromVars<JSONValue>,
): [q: string, vars: string[]] {
  if (typeof obj === "function") {
    const varName = obj(null as unknown as Record<string, unknown>);
    // assert typeof varName === "string"
    return [`$${varName}`, [varName as string]];
  }

  if (Array.isArray(obj)) {
    const [qs, vars] = unzip(obj.map(stringifyQL));
    return [`[${qs.join(", ")}]`, vars.flat(1)];
  }

  if (typeof obj === "object" && obj != null) {
    const [qs, vars] = unzip(Object.entries(obj)!.map(([k, v]) => {
      const [q, vars] = stringifyQL(v);
      return [[k, q] as const, vars];
    }));
    return [
      `{${
        qs.map(([k, v]) => `${k}: ${v}`)
          .join(", ")
      }}`,
      vars.flat(1),
    ];
  }

  return [JSON.stringify(obj), []];
}

// const createArgNode = (name: string, value: ValueNode): ArgumentNode => {
//   return {
//     kind: Kind.ARGUMENT,
//     name: {
//       kind: Kind.NAME,
//       value: name,
//     },
//     value,
//   };
// };

type Mutable<T> = { -readonly [P in keyof T]: T[P] };

const createField = (
  name: string,
  selectionSet: boolean,
  args: ReadonlyArray<ArgumentNode>,
) => {
  const additionalProps: Mutable<
    Pick<FieldNode, "selectionSet" | "arguments">
  > = {};
  if (selectionSet) {
    additionalProps.selectionSet = {
      kind: Kind.SELECTION_SET,
      selections: [],
    };
  }
  if (args.length > 0) {
    additionalProps.arguments = args;
  }
  return {
    kind: Kind.FIELD,
    name: {
      kind: Kind.NAME,
      value: name,
    },
    ...additionalProps,
  } as FieldNode;
};

const findOrCreateField = (
  selections: Array<FieldNode>,
  name: string,
  selectionSet: boolean,
  args: ReadonlyArray<ArgumentNode>,
) => {
  const found = selections.find((n) => n.name.value === name);
  if (found) {
    // if (Boolean(selectionSet) !== Boolean(found.selectionSet)) {
    //   throw new Error("unexpected result")
    // }
    return found;
  }
  const created = createField(name, selectionSet, args);
  selections.push(created);
  return created;
};

const createTargetField = (
  origPath: readonly string[],
  rootSelections: ReadonlyArray<FieldNode>,
  selectionSet: boolean,
  args: ReadonlyArray<ArgumentNode>,
) => {
  const path = [...origPath];
  const last = path.pop()!;
  if (path.length > 0) {
    const first = path.shift()!;
    let field = findOrCreateField(
      rootSelections as Array<FieldNode>,
      first,
      true,
      [],
    );
    for (const p of path) {
      field = findOrCreateField(
        field.selectionSet!.selections as Array<FieldNode>,
        p,
        true,
        [],
      );
    }
    return findOrCreateField(
      field.selectionSet!.selections as Array<FieldNode>,
      last,
      selectionSet,
      args,
    );
  }
  return findOrCreateField(
    rootSelections as Array<FieldNode>,
    last,
    selectionSet,
    args,
  );
};

const createVarDef = (name: string, type: string): VariableDefinitionNode => {
  return {
    kind: Kind.VARIABLE_DEFINITION,
    variable: {
      kind: Kind.VARIABLE,
      name: {
        kind: Kind.NAME,
        value: name,
      },
    },
    type: parseType(type, { noLocation: true }),
  };
};

export interface RebuildQueryParam {
  stages: ComputeStage[];
  renames: Record<string, string>;
}

export interface RebuiltGraphQuery {
  selections: Array<FieldNode>;
  vars: Array<VariableDefinitionNode>;
}

export function rebuildGraphQuery(
  { stages, renames }: RebuildQueryParam,
): RebuiltGraphQuery {
  // let query = "";
  const rootSelections: Array<FieldNode> = [];
  const forwaredVars: Array<VariableDefinitionNode> = [];
  // const forwardedVars: Record<string, string> = {};
  const forwardVar = (name: string) => {
    // if (!Object.hasOwnProperty.call(forwardedVars, name)) {
    //   forwardedVars[name] = stages[0].varType(name);
    // }
    if (!forwaredVars.find((varDef) => varDef.variable.name.value === name)) {
      forwaredVars.push(createVarDef(name, stages[0].varType(name)));
    }
  };
  console.log("stages", stages.map((p) => p.props.path));
  // if (stages.length <= 3) throw new Error("End of life");
  iterParentStages(stages, (stage, children) => {
    const field = stage.props.path[stage.props.path.length - 1];
    // console.log("stages", stages.map((s) => s.props.path));
    // console.log("stage", stage.props);
    // console.log("children", children.map((c) => c.props.path));
    const path = stage.props.materializer?.data["path"] as string[] ?? [field];
    ensure(path.length > 0, "unexpeced empty path");
    // console.log("input type", stage.props.inpType, stage.props.argumentNodes);
    // const args = Object.entries(stage.props.args).map(
    //   ([argName, argValue]) => {
    //     const val = argValue({}, null, {}); // as FromVars<_>
    //     if (typeof val === "function") {
    //       const varName = val(null) as string;
    //       forwardVar(varName);
    //       return createArgNode(argName, {
    //         kind: Kind.VARIABLE,
    //         name: {
    //           kind: Kind.NAME,
    //           value: varName,
    //         },
    //       });
    //     }
    //     console.log("args", stage.props.args);
    //     const [q, vs] = stringifyQL(val as JSONValue);
    //     if (vs.length > 0) {
    //       console.log({ val, q, vs });
    //       throw new Error("not implemented");
    //     }
    //     return createArgNode(argName, GraphQL.parseValue(q));
    //   },
    // );
    const targetField = createTargetField(
      path,
      rootSelections,
      children.length > 0,
      stage.props.argumentNodes ?? [],
    );

    if (stage.props.argumentNodes) {
      for (const argNode of stage.props.argumentNodes) {
        GraphQL.visit(argNode, {
          [Kind.VARIABLE]: (node) => {
            forwardVar(node.name.value);
          },
        });
      }
    }

    // children
    if (children.length > 0) {
      const { selections, vars } = rebuildGraphQuery({
        stages: children,
        renames,
      });
      (targetField.selectionSet!.selections as Array<SelectionNode>).push(
        ...selections,
      );
      forwaredVars.push(...vars);
    }
  });
  return { selections: rootSelections, vars: forwaredVars };
}
