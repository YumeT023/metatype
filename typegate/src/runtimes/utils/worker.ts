import { getLogger } from "../../log.ts";
import { Task, TaskExec } from "./codes.ts";

const logger = getLogger("worker");
logger.info("start webworker");

type TaskModule = Record<string, TaskExec>;

const fns: Record<number, TaskExec> = {};

self.onmessage = async (evt: MessageEvent<Task>) => {
  switch (evt.data.type) {
    case "import_func": {
      const { id, module: path, name, args, context } = evt.data;
      logger.info(`running task ${name} in worker...`);
      const mod: TaskModule = await import(path);

      logger.info(`[${id}] exec func "${name}" from module ${path}`);
      const ret = await mod[name](args, context);
      self.postMessage({ id, data: ret });
      break;
    }

    case "func": {
      const { id, fnId, code, args, context } = evt.data;
      if (code != undefined) {
        fns[fnId] = new Function(`"use strict"; return ${code}`)();
      }

      logger.info(`[${id}] exec func "${fnId}"`);
      const ret = await fns[fnId](args, context);
      self.postMessage({ id, data: ret });
      break;
    }

    default:
      throw new Error(
        `unsupported task type "${(evt.data as any).type as string}"`,
      );
  }
};
