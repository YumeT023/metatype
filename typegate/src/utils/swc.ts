// Copyright Metatype under the Elastic License 2.0.

import { transform } from "swc";
import type { Config } from "swc/types";
import type { TypeGraphDS } from "../typegraph.ts";
import { ensure } from "../utils.ts";

const swcConfig: Config = {
  jsc: {
    target: "es2019",
    parser: {
      syntax: "typescript",
    },
  },
};

export function compileCodes(tg: TypeGraphDS) {
  for (const mat of tg.materializers) {
    const runtime = tg.runtimes[mat.runtime];
    if (runtime.name !== "deno" && runtime.name !== "worker") {
      continue;
    }
    try {
      switch (mat.name) {
        case "function": {
          const prefix = "const f = ";
          const compiled =
            transform(`${prefix}${mat.data.fn_expr as string};`, swcConfig)
              .code;

          ensure(
            compiled.startsWith(prefix),
            "invalid prefix for compiled function expression",
          );

          mat.data.fn_expr =
            transform(compiled.slice(prefix.length), swcConfig).code;

          break;
        }

        case "module":
          mat.data.code = transform(mat.data.code as string, swcConfig).code;
          break;
      }
    } catch (e) {
      throw new Error(
        `materializer ${mat.name} in ${runtime.name} failed compilation: ${e}`,
      );
    }
  }
}
