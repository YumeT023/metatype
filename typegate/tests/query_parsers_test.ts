// Copyright Metatype OÃœ under the Elastic License 2.0 (ELv2). See LICENSE.md for usage.

import { test } from "./utils.ts";

test("GraphQL parser", async (t) => {
  const e = await t.pythonFile("typegraphs/graphql.py");
  /**
   * Types from the parsed TypeGraph
   */
  const types = e.tg.tg.types;
  t.assertSnapshot(types);
});
