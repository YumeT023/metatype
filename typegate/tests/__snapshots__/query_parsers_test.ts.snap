export const snapshot = {};

snapshot[`GraphQL parser 1`] = `
[
  {
    config: {},
    policies: [
    ],
    properties: {
      query: 14,
    },
    required: [
    ],
    runtime: 0,
    title: "graphql_namespaces",
    type: "object",
  },
  {
    config: {},
    policies: [
      0,
    ],
    properties: {
      find: 2,
      profile: 8,
      update: 7,
    },
    required: [
    ],
    runtime: 0,
    title: "user_namespace",
    type: "object",
  },
  {
    config: {},
    input: 3,
    materializer: 1,
    output: 5,
    policies: [
    ],
    rate_calls: false,
    rate_weight: null,
    runtime: 1,
    title: "function_7",
    type: "function",
  },
  {
    config: {},
    policies: [
    ],
    properties: {
      id: 4,
    },
    required: [
    ],
    runtime: 1,
    title: "object_6",
    type: "object",
  },
  {
    config: {},
    format: "uuid",
    policies: [
    ],
    runtime: 1,
    title: "user_id",
    type: "string",
  },
  {
    config: {},
    policies: [
    ],
    properties: {
      id: 4,
      name: 6,
    },
    required: [
    ],
    runtime: 1,
    title: "user_model",
    type: "object",
  },
  {
    config: {},
    policies: [
    ],
    runtime: 1,
    title: "string_2",
    type: "string",
  },
  {
    config: {},
    input: 5,
    materializer: 2,
    output: 5,
    policies: [
    ],
    rate_calls: false,
    rate_weight: null,
    runtime: 1,
    title: "function_8",
    type: "function",
  },
  {
    config: {},
    policies: [
    ],
    properties: {
      picture: 9,
      setPicture: 13,
    },
    required: [
    ],
    runtime: 0,
    title: "profile_namespace",
    type: "object",
  },
  {
    config: {},
    input: 10,
    materializer: 1,
    output: 11,
    policies: [
    ],
    rate_calls: false,
    rate_weight: null,
    runtime: 1,
    title: "function_10",
    type: "function",
  },
  {
    config: {},
    policies: [
    ],
    properties: {
      id: 4,
    },
    required: [
    ],
    runtime: 1,
    title: "object_9",
    type: "object",
  },
  {
    config: {},
    policies: [
    ],
    properties: {
      id: 4,
      url: 12,
    },
    required: [
    ],
    runtime: 1,
    title: "picture_model",
    type: "object",
  },
  {
    config: {},
    format: "uri",
    policies: [
    ],
    runtime: 1,
    title: "string_4",
    type: "string",
  },
  {
    config: {},
    input: 11,
    materializer: 2,
    output: 11,
    policies: [
    ],
    rate_calls: false,
    rate_weight: null,
    runtime: 1,
    title: "function_11",
    type: "function",
  },
  {
    policies: [
    ],
    properties: {
      user: 1,
    },
    runtime: -1,
    title: "Query",
    type: "object",
  },
]
`;
